### PR TITLE
Restructure the `org-redirect` guard to be Angular 17+ compliant

### PR DESCRIPTION
**Pull request by addisonbeck**
# Restructure the `org-redirect` guard to be Angular 17+ compliant

## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/AC-2756

## 📔 Objective

In the web vault: in order to support Angular versions 17 and higher we must
restructure the way organization landing page URL redirects are written and
configured.

Currently most URL guards in the Bitwarden Angular clients are written as
router guard classes. The base `CanActivate` / `CanDeactivate` classes have
been deprecated by the Angular team and are subject to removal. See [the](https://angular.io/guide/deprecations#router-class-and-injectiontoken-guards-and-resolvers)
[Angular deprecation notice](https://angular.io/guide/deprecations#router-class-and-injectiontoken-guards-and-resolvers) for details.

At the time of writing, platform team has already completed this work for
their guards [here](https://bitwarden.atlassian.net/browse/PM-4349). Any guard declarations are swapped from exporting a
class to exporting a function, and the function is added to CanActivate
arrays in routers the same way the classes already are. There are some other
guards that are already function based scattered throughout the codebase that
can also be used as examples for style, like [auth team's lock guard](https://github.com/bitwarden/clients/blob/1aaa88a64d4d48aa75868980eb4c280b672057c3/libs/angular/src/auth/guards/lock.guard.ts).

This PR achieves this goal for the `org-redirect` guard.

This change should result in no functional changes to the system. It is
purely a code restructure to comply with a deprecation in Angular.
This type of redirect occurs in three known places:
```
<https://THE_WEB_VAULT/organizations/{id}>
<https://THE_WEB_VAULT/organizations/{id}/settings>
<https://THE_WEB_VAULT/organizations/{id}/reporting>
```
In each case these URLs redirect to a default landing page that is different
based on the active user's permission setup. Users should be redirected to
whatever submenu is first available for them in the list in the UI. 

### Admin Console Landing Page 

In this example one user has access to manage groups, and the other has access to manage policies.

https://github.com/bitwarden/clients/assets/15897251/27d07cfe-5c86-4837-9651-59594f143cd7

### Admin Console Settings Page

In this example one user is an owner, and so can access billing information, and the second user is an admin, and can not.

https://github.com/bitwarden/clients/assets/15897251/029b389e-4986-4fb9-b550-74df83ed24ff

### Admin Console Reports Page

In this example one user has access to event logs, and the other has access to reports.

https://github.com/bitwarden/clients/assets/15897251/912e8482-7784-4276-a4ab-a87e840f4626

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: apps/web/src/app/admin-console/organizations/guards/org-redirect.guard.ts [hunk 1]
```diff
@@ -1,35 +1,43 @@
...
-    private organizationService: OrganizationService,
-  ) {}
+/**
+ *
+ * `CanActivateFn` that returns a URL Tree redirecting to a caller provided
+ * sub route of `/organizations/{id}/`. If no sub route is provided the URL
+ * tree returned will redirect to `/organizations/{id}` if possible, or `/` if
+ * the user does not have permission to access `organizations/{id}`.
+ */
+export function organizationRedirectGuard(): CanActivateFn {
```
Comment (eliykat) [inline apps/web/src/app/admin-console/organizations/guards/org-redirect.guard.ts:21]: 🎨 (non-blocking): Now that this is a function, you could replace the non-type safe `route.data` property with an explicitly typed parameter:

```ts
export function organizationRedirectGuard(autoRedirectCallback: org => string): CanActivateFn {
```

which I think makes full use of the new "all guards are just functions' design.

**Review by eliykat — DISMISSED** [review]
Great work, I like the tests particularly 🎉

File: apps/web/src/app/admin-console/organizations/guards/org-redirect.guard.ts (hunk 1 above)
Comment (addisonbeck) [reply apps/web/src/app/admin-console/organizations/guards/org-redirect.guard.ts]: Good idea. Implemented as suggested on https://github.com/bitwarden/clients/pull/9552/commits/42f77e10322241ff3bb1b1fc916ee76f9142f041

File: apps/web/src/app/admin-console/organizations/settings/organization-settings-routing.module.ts [hunk 2]
```diff
@@ -20,7 +20,7 @@ const routes: Routes = [
       {
         path: "",
         pathMatch: "full",
-        canActivate: [OrganizationRedirectGuard],
+        canActivate: [organizationRedirectGuard()],
         data: {
           autoRedirectCallback: getSettingsRoute,
         },
```
Comment (eliykat) [inline apps/web/src/app/admin-console/organizations/settings/organization-settings-routing.module.ts:26]: Missed this one

**Review by eliykat — CHANGES_REQUESTED** [review]

File: apps/web/src/app/admin-console/organizations/settings/organization-settings-routing.module.ts (hunk 2 above)
Comment (addisonbeck) [reply apps/web/src/app/admin-console/organizations/settings/organization-settings-routing.module.ts]: Thank you. Added on https://github.com/bitwarden/clients/pull/9552/commits/df39b9c983e958c45380b92848bb51336c5890aa

**Review by eliykat — DISMISSED** [review]

**Review by eliykat — APPROVED** [review]

**Pull request merged**